### ISSUE TITLE
Gromacs Export: Residue name affects molecule type but not atom lines

### ISSUE THREAD

**Issue by megosato** [openforcefield/openff-interchange#642]
**Description**

I’m trying to create a `.gro` and `.top` file using `openff-2.0.0.offxml`  and Interchange. The files output are correct except that I’m trying to name my ligand 3 letter code something other than `UNK` such as `LIG` , but whenever I run `.to_top()`  or `.to_gro()` the ligand gets renamed back to `UNK`. 

**Reproduction**

Please include a [minimally reproducing example](https://stackoverflow.com/help/minimal-reproducible-example) of this bug.

Input file: [lig_41.mol2.zip](https://github.com/openforcefield/openff-interchange/files/11161999/lig_41.mol2.zip)

```
from openff.toolkit import Molecule
from openff.toolkit.typing.engines.smirnoff import ForceField
from openff.interchange import Interchange

lig_mol = Molecule("lig_41.mol2")
lig_mol.name = "LIG"

for atom in lig_mol.atoms:
    atom.metadata["residue_name"] = "LIG"

sage = ForceField('openff-2.0.0.offxml')

interchange = Interchange.from_smirnoff(
    force_field=sage, topology=lig_mol.to_topology()
)

interchange.to_gro("lig_41.gro")
interchange.to_top("lig_41.top")
```

**Output**

Please include the output, including full tracebacks of any errors, resulting from the reproduction.

The final file contained UNK instead of LIG

`lig_41.gro` file:
```
Generated by Interchange
46
    0UNK     C1    1  -3.823  -4.476   1.706
    0UNK     C2    2  -3.958  -4.456   1.733
    0UNK     C3    3  -3.797  -4.579   1.924
    0UNK     C4    4  -3.279  -4.499   1.627
    0UNK     C5    5  -3.232  -4.696   1.803
    0UNK     C6    6  -3.933  -4.559   1.954
    0UNK     C7    7  -3.742  -4.537   1.802
```

`lig_41.top` file:
```
[ moleculetype ]
; Name	nrexcl
LIG	3

[ atoms ]
;num, type, resnum, resname, atomname, cgnr, q, m
     1 C1                      0 UNK      C1            1        -0.11459000        12.01078000
     2 C2                      0 UNK      C2            2        -0.11648000        12.01078000
     3 C3                      0 UNK      C3            3        -0.15263000        12.01078000
     4 C4                      0 UNK      C4            4         0.38417000        12.01078000
     5 C5                      0 UNK      C5            5         0.27586001        12.01078000
     6 C6                      0 UNK      C6            6        -0.15433000        12.01078000
     7 C7                      0 UNK      C7            7         0.08815000        12.01078000
```

**Software versions**

- Which operating system and version are you using?
Mac OS Ventura 13.1 (M1 Mac)

- How did you install Interchange?
`conda create -n interchange -c conda-forge openff-interchange`

and also `pip install -e .` from main

- What is the output of running `conda list`?
Using conda:
```
openff-amber-ff-ports     0.0.3              pyh6c4a22f_0    conda-forge
openff-forcefields        2.0.0              pyh6c4a22f_0    conda-forge
openff-interchange        0.2.3              pyhd8ed1ab_1    conda-forge
openff-interchange-base   0.2.3              pyhd8ed1ab_1    conda-forge
openff-toolkit            0.12.1             pyhd8ed1ab_1    conda-forge
openff-toolkit-base       0.12.1             pyhd8ed1ab_1    conda-forge
openff-units              0.2.0              pyh1a96a4e_0    conda-forge
openff-utilities          0.1.7              pyh1a96a4e_0    conda-forge
```

Using pip install:
```
openff-amber-ff-ports     0.0.3              pyh6c4a22f_0    conda-forge
openff-forcefields        2023.02.1          pyh1a96a4e_0    conda-forge
openff-interchange        0.3.0b4+28.g102ce1a9          pypi_0    pypi
openff-models             0.0.4              pyh1a96a4e_0    conda-forge
openff-toolkit            0.12.1             pyhd8ed1ab_1    conda-forge
openff-toolkit-base       0.12.1             pyhd8ed1ab_1    conda-forge
openff-units              0.2.0              pyh1a96a4e_0    conda-forge
openff-utilities          0.1.8              pyh1a96a4e_0    conda-forge
```

**Comment by mattwthompson** [comment]
Nice catch - I think this is valid, but it's also tricky to get right in a way that doesn't interfere with biopolymers. 
The direct change should be straightforward enough but the logic around assigning GROMACS molecule names needs some more work since OpenFF tracks molecules and residues but GROMACS. Strictly speaking those three-letter codes are _residue_ names (in [GRO](https://manual.gromacs.org/documentation/2023/reference-manual/topologies/topology-file-formats.html#coordinate-file) and [TOP/ITP](https://manual.gromacs.org/documentation/2023/reference-manual/topologies/topology-file-formats.html#molecule-itp-file) files) and in preparing this `Topology` no residue information was set, so Interchange just writes out the `UNK`s you see. Setting the residue information in preparation is a possible solution, but not the greatest (related https://github.com/openforcefield/openff-interchange/issues/594 https://github.com/openforcefield/openff-toolkit/issues/1554). A more involved solution would set residue names to match `Molecule` names, but this will take some care to think about which cases it can apply to. Doing this when no atoms in a `Topology` have residue information set but its `Molecule`s are named is what comes to mind first; this should work for your systems and most small-molecule systems without interfering with well-prepared biopolymer systems. This might also work for protein-ligand systems when the ligand has a name but its atoms don't have residue names, but more tests will be needed.

When I wrote the GRO writer, I wanted it to work with a minimum amount of information (positions, elements, box, and not much else), more than what happens when calling `Interchange.from_smirnoff`. But then I didn't go back and add in the extra information if it's available.

I don't think this a critical bug so I won't backport it to the 0.2.x line. In lieu of #635 and some upcoming changes, the fix for this will probably end up in 0.3.1.

**Comment by mattwthompson** [comment]
I think I fixed this in #646, but there are still some cases left un-tested. Expect it to land in version 0.3.1 in late April.

Please continue to report any questionable behavior you run into - this issue is a **textbook** example of how to report bugs (and make my job easy).
